Replace $_GET/$_POST/$_FILES with App::getRequestXxx()

abbadon1334 opened the pull request:
fix #2009

abbadon1334 committed:
Fix PrettyUrl legitimate path

abbadon1334 committed:
Add Unit tests for App::url

abbadon1334 committed:
add default excepted page and extension

abbadon1334 committed:
fix 7.4 array merge with keys / spread operator

abbadon1334 committed:
Add more test cases for App:url

abbadon1334 committed:
Reorganize code

abbadon1334 committed:
Improve converage

abbadon1334 committed:
Improve App::url tests

abbadon1334 committed:
Skip $useRequestUrl=true due to REQUEST_URI

abbadon1334 committed:
Change REQUEST_URI to Psr7 methods

abbadon1334 committed:
Allow $useRequestUrl=true

abbadon1334 committed:
clean up

abbadon1334 committed:
Improve error message

abbadon1334 committed:
CS Fix

abbadon1334 committed:
Merge branch 'develop' into add_tests_app_url

abbadon1334 committed:
Add App::$urlBuildingPage to customize url building page

abbadon1334 committed:
Comment out App::url failing tests

abbadon1334 committed:
Merge branch 'develop' into fix_pretty_url_legitimate_path

abbadon1334 committed:
Avoid change of global var $_SERVER

abbadon1334 committed:
Fix missing request in App for Unit tests

abbadon1334 committed:
Comment out not working tests

abbadon1334 committed:
removed getRequestUrl

abbadon1334 committed:
Add failing tests

abbadon1334 committed:
Add methods App::issetRequestXxx App::getRequestXxx

abbadon1334 committed:
Replaced all $_FILES

abbadon1334 committed:
Replace $_GET App::getRequestGetParam

abbadon1334 committed:
Replace $_GET App::getRequestGetParam

mvorisek commented on src/App.php:
```diff
@@ -1193,4 +1224,82 @@ private function getRenderedPortals(): array
+     */
+    public function issetRequestGetParam(string $key): bool
+    {
+        return isset($this->request->getQueryParams()[$key]);
```
will return false when null

mvorisek commented on src/App.php:
```diff
@@ -1193,4 +1224,82 @@ private function getRenderedPortals(): array
+    /**
+     * Return true if $_GET[$key] exists.
+     */
+    public function issetRequestGetParam(string $key): bool
```
`hasXxx` name LBTM

mvorisek commented on src/App.php:
```diff
@@ -1193,4 +1224,82 @@ private function getRenderedPortals(): array
+     */
+    public function getRequestGetParam(string $key)
+    {
+        return $this->getRequestGetParams()[$key] ?? null;
```
with https://github.com/atk4/ui/pull/2101/files#r1316302660 returning `true` for `null`, this should throw if the key is not present

mvorisek commented on src/App.php:
```diff
@@ -1193,4 +1224,82 @@ private function getRenderedPortals(): array
+    /**
+     * Return $_GET param by key or null if not exists.
+     *
+     * @return mixed
```
AFAK the GET can be string, null

and array of strings, but we do not want to support it probably using this method, as longer alternative exists :)

mvorisek replied:
test cases:

```
var_dump($app->hasRequestGetParam('x')); // test URL: ?x, passed, empty query value (without '=') is still non-null/string
var_dump($app->getRequestGetParam('x')); // test URL: ?x[]=1&x[]=2, TypeError: Return value of Atk4\Ui\App::getRequestGetParam() must be of the type string, array returned
```

abbadon1334 committed:
Replace $_GET App::getRequestGetParam

abbadon1334 committed:
Add App::tryGetRequestGetParam not raise exceptiona nd return null if not present

abbadon1334 committed:
Replace $_GET App::has/get/tryGet/RequestGetParam

abbadon1334 committed:
Replace $_GET/$_POST App::has/get/tryGet/RequestGet/PostParam

abbadon1334 committed:
Replace $_GET/$_POST App::has/get/tryGet/RequestGet/PostParam

abbadon1334 committed:
Cs Fix

abbadon1334 committed:
Fix Url building if path starts with ..

abbadon1334 committed:
Fix JSON POST test

abbadon1334 committed:
Fix stan

abbadon1334 committed:
Raise composer require to nyholm/psr7 ( failing lowest tests)

mvorisek commented on composer.json:
```diff
@@ -53,7 +53,7 @@
         "php": ">=7.4 <8.3",
         "atk4/data": "dev-develop",
-        "nyholm/psr7": "^1.4",
+        "nyholm/psr7": "^1.6.0",
```
```suggestion
        "nyholm/psr7": "^1.6",
```

looks fine - https://github.com/Nyholm/psr7/blob/1.6.0/CHANGELOG.md

make sure you change it also in the `require-release`!

abbadon1334 committed:
Align composer require-release to require

abbadon1334 committed:
Cs Fix

abbadon1334 committed:
Revert "Cs Fix"

This reverts commit 4a9c7114489dfc1f422dfff01d2b4878ee89cc06.

abbadon1334 committed:
Cs Fix with latest version

abbadon1334 committed:
Strong type for $urlBuildingXxx vars

abbadon1334 committed:
Use CreateApp

abbadon1334 committed:
Remove message from Assert

abbadon1334 committed:
Cs Fix

abbadon1334 committed:
Raise Nyholm/psr7 from 1.4 to 1.6

abbadon1334 committed:
Merge branch 'develop' into use_request_not_globals

mvorisek committed:
Merge branch 'develop' into add_tests_app_url

mvorisek committed:
DEBUG REVERT ALL

mvorisek committed:
drop protected App::getRequestUrl() method

mvorisek committed:
introduce App::$urlBuildingIndexPage prop

mvorisek committed:
replace $_SERVER['REQUEST_URI'] with PSR7 request

mvorisek committed:
fix cs

mvorisek committed:
original AppTest changes

mvorisek committed:
no manual assert context dump

mvorisek committed:
allow to create test App with defaults

mvorisek committed:
always use test App in tests

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
Merge PR 2096

mvorisek committed:
remaining reverted changes from PR 2096

mvorisek committed:
Merge develop

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
Merge branch 'fix_pretty_url_legitimate_path'

mvorisek commented on demos/_unit-test/post.php:
```diff
@@ -10,6 +10,14 @@
 /** @var \Atk4\Ui\App $app */
 require_once __DIR__ . '/../init-app.php';
 
+// Creation of request from globals depenmds on getallheaders
```
no hacks in individual files, this must be fixed in central place

if the comment about `getallheaders` is true, then https://github.com/Nyholm/psr7-server/blob/1.0.2/src/ServerRequestCreator.php#L52 needs to be replaced

abbadon1334 replied:
i move it to demos/init-app.php

i don't think the internal creation of ServerRequest must be changed, we need it only in unit tests running via CLI

mvorisek replied:
discussed in https://github.com/atk4/ui/pull/2101#discussion_r1321152445

mvorisek commented on src/App.php:
```diff
@@ -1151,4 +1182,101 @@ private function outputResponseJson($data): void
+    /**
+     * Return true if $_GET[$key] exists.
+     */
+    public function hasRequestGetParam(string $key): bool
```
It seems you completely missed docs, they cannot be checked using CI, so for every change you commit, you must keep them in sync!

abbadon1334 replied:
you are right, unit test failing makes me do some too fast cut/paste.
I update them

abbadon1334 committed:
Add correct DocBlock

abbadon1334 committed:
Add method App::tryGetRequestUploadedFile for consistency

abbadon1334 committed:
Add to _SERVER key CONTENT_TYPE
Running via CLI $_POST will be created on if content-type= application/x-www-form-urlencoded or multipart/form-data

abbadon1334 committed:
Removed Hack

abbadon1334 committed:
Cs Fix

abbadon1334 committed:
Fix Typo

abbadon1334 committed:
Fix getRequestPostParam and tryGetRequestPostParam cases

abbadon1334 committed:
Fix getRequestPostParam and tryGetRequestPostParam cases

abbadon1334 committed:
Recheck usage of Try with useless coalesce

mvorisek commented on tests/DemosTest.php:
```diff
@@ -502,7 +503,12 @@ public function provideDemoAssertJsonResponsePostCases(): iterable
-        $response = $this->getResponseFromRequest($path, ['form_params' => $postData]);
+        $response = $this->getResponseFromRequest($path, [
+            'headers' => [
+                'Content-Type' => 'application/x-www-form-urlencoded',
```
if `$postData` is an array (and in our case always is), the content-type should not be needed to be set explicitly - can this be fixed in the guzzle lib directly?

abbadon1334 replied:
From ServerRequestCreator

_POST must be loaded only if `Content-Type` equals to : `application/x-www-form-urlencoded` OR `multipart/form-data`

```php
    public function fromGlobals(): ServerRequestInterface
    {
        $server = $_SERVER;
        if (false === isset($server['REQUEST_METHOD'])) {
            $server['REQUEST_METHOD'] = 'GET';
        }

        $headers = \function_exists('getallheaders') ? getallheaders() : static::getHeadersFromServer($_SERVER);

        $post = null;
        if ('POST' === $this->getMethodFromEnv($server)) {
            foreach ($headers as $headerName => $headerValue) {
                if (true === \is_int($headerName) || 'content-type' !== \strtolower($headerName)) {
                    continue;
                }
                if (\in_array(
                    \strtolower(\trim(\explode(';', $headerValue, 2)[0])),
                    ['application/x-www-form-urlencoded', 'multipart/form-data']
                )) {
                    $post = $_POST;

                    break;
                }
            }
        }

        return $this->fromArrays($server, $headers, $_COOKIE, $_GET, $post, $_FILES, \fopen('php://input', 'r') ?: null);
    }
```

In Browser, this is done internally.
In Guzzle probably leave to the user all options, even bad formatted requests.
I think We can add Content-Type if POST refactoring it in our code: https://github.com/atk4/ui/blob/d64706e9f8cf3eab4b1b64cadea8eb1b6881774e/tests/DemosTest.php#L199

mvorisek replied:
Yes, the reason is https://github.com/Nyholm/psr7-server/blob/1.0.2/src/ServerRequestCreator.php#L62 - next time, please post a link instead of copy pasting the code here... Anyway, see https://github.com/atk4/ui/commit/6143477f79da6de711472bae89ffaeae48751ced, when all request headers are reconstructed into $_SERVER, no additional code is needed

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
Merge branch 'fix_pretty_url_legitimate_path'

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
silence failing tests

mvorisek committed:
enable commented tests

mvorisek committed:
add wrongly removed check_unset_page test

mvorisek committed:
Merge branch 'fix_pretty_url_legitimate_path'

mvorisek committed:
REVERT ALL (reset to develop)

mvorisek commented on src/App.php:
```diff
@@ -617,7 +618,7 @@ protected function createRequestPathFromLocalPath(string $localPath): string
             } else {
                 $request = \Symfony\Component\HttpFoundation\Request::createFromGlobals();
```
TODO for myself - this relies on $_GET, rework it in https://github.com/atk4/ui/pull/2095 and test by unsetting `$_GET`

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
unmerge PR 2095 ('fix_pretty_url_legitimate_path' branch)

mvorisek committed:
REVERT ALL (reset to develop)

mvorisek committed:
Merge branch 'develop' into use_request_not_globals

mvorisek commented on src/App.php:
```diff
@@ -26,6 +26,7 @@
 use Psr\Http\Message\ResponseInterface;
```
@abbadon1334 please explain this coverage change:

![image](https://github.com/atk4/ui/assets/2228672/5f3fe01e-f422-4887-a3f7-c4337ecdcb34)

- it seems some error is newly triggered - how to repro it?
- it should NOT be triggered by these refactoring changes
- a Behat test to actually trigger it will be good, as a separate PR please

mvorisek commented on src/App.php:
```diff
@@ -300,8 +301,8 @@ public function caughtException(\Throwable $exception): void
 
-        if (($this->isJsUrlRequest() || strtolower($_SERVER['HTTP_X_REQUESTED_WITH'] ?? '') === 'xmlhttprequest')
-                && !isset($_GET['__atk_tab'])) {
+        if (($this->isJsUrlRequest() || $this->request->getHeaderLine('HTTP_X_REQUESTED_WITH') === 'xmlhttprequest')
```
![image](https://github.com/atk4/ui/assets/2228672/5f3fe01e-f422-4887-a3f7-c4337ecdcb34)

the removed `strtolower` caused 500 HTTP code on AJAX failure to be returned instead of 200

I will fix this, headers are case sensitive, so no `strtolower` is wanted but instead compare to `XMLHttpRequest` value

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
replace all 'isset\(\$_GET\[(.+?)\]' matches

mvorisek committed:
replace all '\$_GET\[(.+?)\]' matches

mvorisek committed:
replace all '\$_POST\[(.+?)\](?= \?\?)' matches

mvorisek committed:
replace all '\$_POST\[(.+?)\]' matches

mvorisek committed:
rm CreateAppTrait::setGlobalsFromRequest() test method

mvorisek committed:
Merge branch 'develop'

mvorisek committed:
replace all '\$_GET\[(.+?)\](?= \?\?)' matches

mvorisek committed:
replace remaining $_GET/$_POST

mvorisek committed:
replace/emulate $_FILES

mvorisek committed:
fix $_SERVER emulation for POST test

mvorisek commented on src/App.php:
```diff
@@ -320,6 +321,98 @@ public function getRequest(): ServerRequestInterface
+    /**
+     * Get the value of a specific GET parameter from the HTTP request.
+     */
+    public function getRequestGetParam(string $key): string
```
@abbadon1334 do you prefer `getRequestGetParam` over `getRequestQueryParam`?

or what about `getRequestQuery($k)`/`getRequestPost($k)`?

abbadon1334 replied:
laravel: `$request->query('name');`
slim: `$request->getQueryParam('page');`
Symfony: `$request->query->get('name');`

i think **Query** is the correct way of naming it, but i should leave **Param**.
So i think the best option is: `getRequestQueryParam($k)` and `getRequestPostParam($k)`

WDYT?

mvorisek replied:
I agree. Renamed. `->getApp()->getRequestQueryParam()` seems however quite long, as I expect it to be used quite frequently, we can introduce shortcut method in View.

mvorisek committed:
unset $_GET, $_POST, ... after App is created in demos/testing

mvorisek committed:
add App request shortcut methods

mvorisek committed:
add edge cases empty GET param test

mvorisek committed:
test all exceptions

mvorisek commented on demos/init-app.php:
```diff
@@ -29,6 +29,12 @@
+unset($_GET);
+unset($_POST);
+unset($_FILES);
+unset($_COOKIE);
```
@abbadon1334 one last thing to address, this unset is breaking one Behat test

abbadon1334 replied:
Am i wrong, or this is needed only for PHPUnit for continuous/fast calls without isolation of requests?

mvorisek replied:
The `unset($_XXX)` there is for non-phpunit actually, it is there to make sure every atk4/ui component use the request details from App/request object. See the Behat failure, it can be easily reproduced localy /w any browser, if these unsets are present, the filtering does not work.

mvorisek replied:
please make the tests passing - but please do NOT force push the branch nor merge in develop - I will then do final review and merge

mvorisek committed:
rename "GetParam" to "QueryParam"

mvorisek committed:
fix phpstan

mvorisek committed:
fix $_COOKIE unset in demos

mvorisek commented:
Thank you @abbadon1334, quality PR!
